packages/core/src/amazonq/commons/diff.ts

```diff
@@ -32,3 +34,126 @@ export function createAmazonQUri(path: string, tabId: string) {
+    const code = message.code as string
+    const indentRange = new vscode.Range(new vscode.Position(selection.start.line, 0), selection.active)
+    let indent = doc.getText(indentRange)
+
+    if (indent.trim().length !== 0) {
+        indent = ' '.repeat(indent.length - indent.trimStart().length)
+    }
+
+    let trimmedCode = code
+    if (code.endsWith('\n')) {
+        const startLineText = doc.lineAt(selection.start.line).text
+        if (!startLineText.endsWith('\n')) {
+            trimmedCode = code.trimEnd()
+        }
+    }
+
+    const indentedCode = trimmedCode
+        .split('\n')
+        .map((line) => indent + line)
+        .join('\n')
+
+    return indentedCode
+}
+
+/**
+ * Creates a selection range from the given document and selection.
+ * If a user selects a partial code, this function generates the range from start line to end line.
+ *
+ * @param {vscode.TextDocument} doc - The VSCode document where the selection is applied.
+ * @param {vscode.Selection} selection - The selection range in the document.
+ * @returns {vscode.Range} - The VSCode range object representing the start and end of the selection.
+ */
+export function getSelectionFromRange(doc: vscode.TextDocument, selection: vscode.Selection) {
+    return new vscode.Range(
+        new vscode.Position(selection.start.line, 0),
+        new vscode.Position(selection.end.line, doc.lineAt(selection.end.line).range.end.character)
+    )
+}
+
+/**
+ * Applies the given code to the specified range in the document.
+ * Saves the document after the edit is successfully applied.
+ *
+ * @param {vscode.TextDocument} doc - The VSCode document to which the changes are applied.
+ * @param {vscode.Range} range - The range in the document where the code is replaced.
+ * @param {string} code - The code to be applied to the document.
+ * @returns {Promise<void>} - Resolves when the changes are successfully applied and the document is saved.
+ */
+export async function applyChanges(doc: vscode.TextDocument, range: vscode.Range, code: string) {
+    const edit = new vscode.WorkspaceEdit()
+    edit.replace(doc.uri, range, code)
+    const successfulEdit = await vscode.workspace.applyEdit(edit)
+    if (successfulEdit) {
+        await doc.save()
+    }
+}
+
+/**
+ * Creates a temporary file for diff comparison by cloning the original file
+ * and applying the proposed changes within the selected range.
+ *
+ * @param {vscode.Uri} originalFileUri - The URI of the original file.
+ * @param {any} message - The message object containing the proposed code changes.
+ * @param {vscode.Selection} selection - The selection range in the document where the changes are applied.
+ * @returns {Promise<vscode.Uri>} - A promise that resolves to the URI of the temporary file with the applied changes.
+ */
+export async function createTempFileForDiff(originalFileUri: vscode.Uri, message: any, selection: vscode.Selection) {
```

Comment:
even this could live in core, and probably just `packages/core/src/shared/utilities/textDocumentUtilities.ts`